grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/fragment/Fragments.java

```diff
@@ -191,4 +167,33 @@ static void applyTypeLabelsToTraversal(
             traversal.has(property.name(), P.within(typeIds));
         });
     }
+
+    /**
+     * Optionally traverse from a shortcut edge to the role-type it mentions, plus any super-types.
+     *
+     * @param traversal the traversal, starting from the shortcut edge
+     * @param roleType the variable to assign to the role-type. If not present, do nothing
+     */
+    static void traverseRoleTypeFromShortcutEdge(GraphTraversal<Vertex, Edge> traversal, Optional<Var> roleType) {
+        roleType.ifPresent(var -> {
+            // Access role-type ID from edge
+            Var roleTypeIdProperty = Var.anon();
+            Var edge = Var.anon();
+            traversal.as(edge.getValue()).values(ROLE_TYPE_ID.name()).as(roleTypeIdProperty.getValue());
+
+            // Look up direct role-type using ID
+            Var directRoleType = Var.anon();
+            GraphTraversal<Vertex, Vertex> vertexTraversal =
+                    traversal.V().as(directRoleType.getValue())
```

Comment:
I don't know. I'll investigate - don't think I'll be able to write a test for it though!